keps/sig-node/20190806-memory-manager.md

```diff
@@ -0,0 +1,264 @@
+
+- [Device Manager][device-manager] is a one of features provides topology hint for Topology Manager. The main objectivity of Device Manager is allow vendors to adverise their resources to kubelet.
+
+- [Hugepages][hugepages] is a feature allows container to consume pre-allocated hugepages with pod isolation of huge pages.
+
+- [Node Allocatable Feature][node-allocatable-feature] is a feature helps to reserve compute resources to prevent resource starvation. The kube-reserved and system-reserved is used to reserve resources for kubelet and system(OS, etc). Now following resources are supported to reserve: cpu, memory, ephemral storage.
+
+## Related issues
+
+- [Hardware topology awareness at node level (including NUMA)][numa-issue]
+- [Support Container Isolation of Hugepages][hugepage-issue]
+- [Support isolating memory to single NUMA node][memory-issue]
+
+## Goals
+
+- Guarantee alignment of resources for isolating container's Memory and Hugepages with CPU and I/O devices.
+
+- Provide topology hint for Memory and Hugepages for Topology manager.
+
+## Non-Goals
+- Updating scheduler, pod spec is out of scope at this point.
+
+- This proposal only focus on Linux based system.
+
+## User Stories
+
+### Story 1 : Networking Acceleration using DPDK
+
+- DPDK(Data Plane Development Kit) is set of libraries to accelerate packet processing on userspace. DPDK requires dedicated resources and alignment of resources under NUMA awareness. There should be a way to gurantee reserve Memory and Hugepages with other computing resources on desinated Socket for containerized VNF using DPDK.
+
+### Story 2 : Database
+
+- Databases(Oracle, PostgreSQL, MySQL, MongoDB, SAP) consumes massive Memory and Hugepages, in order to reduce memory access latency and improve performance dramatically resources(CPU, Memory, Hugepages, I/O Devices) should be aligned.
+
+# Proposal
+
+## Proposed Changes
+
+### New Component: Memory Manager
+
+A new component of Kubelet enables NUMA-awareness for memory and hugepages. The main roles of this component are listed below.
+
+
+#### Calculate NUMA node affinity for memory and hugepages
+>NUMA node affinity represents that which NUMA node has enough capacity of memory and/or hugepages for a container. To calculate affinity Memory Manager takes capacity of memory and pre-allocated hugepages per NUMA node except system and Kubenetes reserved capacity by Node Allocatable feature. Then when pod admit is requested, Memory Manager checks resources availablity of each NUMA nodes and reserves resources internally. 
+- Node affinity of memory can be calcuated by below formulas.
+  - Allocatable Memory of NUMA node = Total Memory of NUMA node - Hugepages - system reserved - kubernetes reserved.
+  - Allocatable Memory of NUMA node >= Guaranteed memory for a container.
+
+- Node affinity of hugepage can be calculated by below formulas.
+  - Available huagpages of NUMA node = Total hugepages of NUMA node - reserved by system.
```

Comment:
Yes, you are right, current kubelet doesn`t support to reserve huagpages for system.
I don`t mention about "[update existing hugepages kep](https://github.com/kubernetes/enhancements/pull/1199)" in this KEP, It's my bad.
It includes "Support reserve huge pages for system on Node Allocatable feature" and "Update cAdviser to discover and store hugepages per NUMA node"

It seems that i can put extra description here for other KEP(updating hugepage) here or do not mention about hugepages that "resered by system" here.
But, I'm not sure which way is better choice.

Reply:
IMHO it not so critical, because once you use the host as Kubernetes node, you should not have an additional application on top of this host that uses huge pages.

Reply:
> 
> 
> IMHO it not so critical, because once you use the host as Kubernetes node, you should not have an additional application on top of this host that uses huge pages.

Yes, in few cases, it is not critical.
Unfortunately ,as I know most modern NFV cluster have DPDK based packet switcher such as OVS-DPDK, VPP or tungsten-fabric on their nodes.

The critical question is most of them are not integrated to kubernetes well, they are just running on bare-metal or host(i mean they are not containerized) and consuming exclusive cpu cores and hugepages of specific NUMA node.

and it will makes the issue which is generating un-reliable topolgy hint from memory manager.
I don't want to face that situation and it was the main reason why I mentioned reserved hugepages here.

Fortunately, We have good news for this issue.
1) we will have `--reserved-cpus` option for kubelet in 1.17. //thanks @jiayingz 
2) we will have reserve flag for hugeapges in 1.18 or later //@odinuge opened [PR](https://github.com/kubernetes/kubernetes) aginst it, in fact it was the part of original hugepages kep, you can see it [here](https://github.com/kubernetes/enhancements/blame/master/keps/sig-node/20190129-hugepages.md#L154).

Reply:
@bg-chun thanks for the information, regarding huge pages reservation, we will have the same problem that we have with memory, how can you know huge pages from what NUMA node the system will use?

Reply:
For hugepages, my idea is adding a new flag to reserve hugepages per numa node.
I think we can extend this approach(https://github.com/kubernetes/kubernetes/pull/83541) later.